apps/onboarding/src/gateway/rules/signature.rule.ts

```diff
@@ -0,0 +1,43 @@
+import { StartSessionDto } from '@app/onboarding/dto/StartSessionDto'
+import { Err, Ok, RootError } from '@celo/base/lib/result'
+import { recoverMessageSigner } from '@celo/contractkit/lib/utils/signing-utils'
+import { hashMessage } from '@celo/utils/lib/signatureUtils'
+import { Injectable } from '@nestjs/common'
+import { Rule, RuleID } from './rule'
+
+export enum SignatureErrorTypes {
+  InvalidSignature = 'InvalidSignature'
+}
+
+export class InvalidSignature extends RootError<SignatureErrorTypes> {
+  constructor() {
+    super(SignatureErrorTypes.InvalidSignature)
+  }
+}
+
+@Injectable()
+export class SignatureRule implements Rule<{}, InvalidSignature> {
+  getID() {
+    return RuleID.Signature
+  }
+
+  async verify(startSessionDto: StartSessionDto, config, context) {
+    const signature = startSessionDto.signature
+    const account = startSessionDto.externalAccount
+    const message = hashMessage(`komenci:login:${account}`)
+    const signer = recoverMessageSigner(message, signature)
+    if (signer.toLocaleLowerCase() === account.toLocaleLowerCase()) {
```

Comment:
Can we switch the check around so that this returns `Err(new InvalidSignature())` if they don't match? This way if someone adds a new check in the future, they can add it after this check without needing to restructure it.